lib/internal/url.js

```diff
@@ -1500,38 +1500,85 @@ function fileURLToPath(path, options = kEmptyObject) {
+// @see https://www.ietf.org/rfc/rfc1738.txt 2.2. URL Character Encoding Issues
+// Back-slashes must be special-cased on Windows, where they are treated as path separator.
 const percentRegEx = /%/g;
 const backslashRegEx = /\\/g;
 const newlineRegEx = /\n/g;
 const carriageReturnRegEx = /\r/g;
 const tabRegEx = /\t/g;
-const questionRegex = /\?/g;
+const quoteRegEx = /"/g;
 const hashRegex = /#/g;
+const spaceRegEx = / /g;
+const lowerThanRegEx = /</g;
+const greaterThanRegEx = />/g;
+const questionRegex = /\?/g;
+const openSquareBracketRegEx = /\[/g;
+const closeSquareBracketRegEx = /]/g;
+const openCurlyBracketRegEx = /{/g;
+const closeCurlyBracketRegEx = /}/g;
+const caretRegEx = /\^/g;
+const backtickRegEx = /`/g;
+const verticalBarRegEx = /\|/g;
 
 function encodePathChars(filepath, options = kEmptyObject) {
-  const windows = options?.windows;
-  if (StringPrototypeIndexOf(filepath, '%') !== -1)
+  if (StringPrototypeIncludes(filepath, '%')) {
     filepath = RegExpPrototypeSymbolReplace(percentRegEx, filepath, '%25');
-  // In posix, backslash is a valid character in paths:
-  if (!(windows ?? isWindows) && StringPrototypeIndexOf(filepath, '\\') !== -1)
-    filepath = RegExpPrototypeSymbolReplace(backslashRegEx, filepath, '%5C');
-  if (StringPrototypeIndexOf(filepath, '\n') !== -1)
+  }
+
+  if (StringPrototypeIncludes(filepath, '\t')) {
+    filepath = RegExpPrototypeSymbolReplace(tabRegEx, filepath, '%09');
+  }
+  if (StringPrototypeIncludes(filepath, '\n')) {
     filepath = RegExpPrototypeSymbolReplace(newlineRegEx, filepath, '%0A');
-  if (StringPrototypeIndexOf(filepath, '\r') !== -1)
+  }
+  if (StringPrototypeIncludes(filepath, '\r')) {
     filepath = RegExpPrototypeSymbolReplace(carriageReturnRegEx, filepath, '%0D');
-  if (StringPrototypeIndexOf(filepath, '\t') !== -1)
-    filepath = RegExpPrototypeSymbolReplace(tabRegEx, filepath, '%09');
+  }
+  if (StringPrototypeIncludes(filepath, ' ')) {
+    filepath = RegExpPrototypeSymbolReplace(spaceRegEx, filepath, '%20');
+  }
+  if (StringPrototypeIncludes(filepath, '"')) {
+    filepath = RegExpPrototypeSymbolReplace(quoteRegEx, filepath, '%22');
+  }
+  if (StringPrototypeIncludes(filepath, '#')) {
+    filepath = RegExpPrototypeSymbolReplace(hashRegex, filepath, '%23');
+  }
+  if (StringPrototypeIncludes(filepath, '<')) {
+    filepath = RegExpPrototypeSymbolReplace(lowerThanRegEx, filepath, '%3C');
+  }
+  if (StringPrototypeIncludes(filepath, '>')) {
+    filepath = RegExpPrototypeSymbolReplace(greaterThanRegEx, filepath, '%3E');
+  }
+  if (StringPrototypeIncludes(filepath, '?')) {
+    filepath = RegExpPrototypeSymbolReplace(questionRegex, filepath, '%3F');
+  }
+  if (StringPrototypeIncludes(filepath, '[')) {
+    filepath = RegExpPrototypeSymbolReplace(openSquareBracketRegEx, filepath, '%5B');
+  }
+  if (!options.windows && StringPrototypeIncludes(filepath, '\\')) {
+    filepath = RegExpPrototypeSymbolReplace(backslashRegEx, filepath, '%5C');
+  }
```

Comment:
i think you can use replace with an object map of replacements (in place of a string or a callback function), and do most of this all in one go rather than one at a time?

Reply:
I'm not sure what you are you talking about, can you give an example?

Reply:
ahhh nvm, i'm thinking of JSON.stringify's replacer arg.

altho this would actually also work with a single replacer function and a single regex unconditionally applied, and then a map of replacements?

Reply:
You mean like that: https://github.com/nodejs/node/blob/ef18a5874e019f2552fdcab85ba7703531809a4f/lib/internal/url.js#L1509-L1537

See https://github.com/nodejs/node/pull/54545#issuecomment-2309017414 for the benchmark results